util/editEntry.js

```diff
@@ -2,72 +2,70 @@ import { testSchema } from "../schemas/testSchema";
     const query = { _id: id };
     const newEngagement = { ...validData, _id: id, testPlanId: testplanId };
-    try {
-      // Connect to the Database
-      var db = await connectToDb();
-    } catch (err) {
-      return { statusCode: 500, message: "Unable to connect to Mongo Database Server", info: db, error: err }
-    }
     try {
       // Update the Database with new Engagement
       var QueryResult = await db.collection("engagements").replaceOne(query, newEngagement);
     } catch (err) {
-      // Mongo-Side Validation failure should occur here
-      return { statusCode: 400, message: "Mongo Database Query was unable to Validate or otherwise Failed", info: QueryResult, error: err }
+      // Mongo-Side Validation failures should occur here
+      return { statusCode: 400, message: "MongoDB Query Failed or could not Validate", info: QueryResult, errorName: err.name, error: err.message }
     }
     // Edit was Successful!
     return { statusCode: 200, message: "Success" }
   } else {
-    // Schema is invalid or invalid Ids
-    let responseText = "Validation Failed: ";
-    if (!valid && !validObjectIds) {
-      responseText += "Contains invalid MongoId and Incorrectly formatted data"
-    }
-    else if (!validObjectIds) {
-      responseText += "Contains invalid MongoId"
-    }
-    else {
-      responseText += "Incorrectly formatted data"
-    }
-    return { statusCode: 422, message: responseText }
+    // Invalid Ids
+    return { statusCode: 422, message: "Validation Failed: Contains invalid MongoId(s)" }
   }
 }
 
 export async function editTestPlan(data) {
   try {
-    // Check that data is formatted correctly
-    var valid = await testPlanSchema.isValid(data);
-    // Check that all Id strings are Valid Mongo Object Ids
-    var validDevices = !data.summaryBOM.map((dev) => ObjectId.isValid(dev.deviceId)).includes(false);
-    var validTestCases = !data.testCases.map((str) => ObjectId.isValid(str)).includes(false);
-    var validObjectIds = validDevices && validTestCases && ObjectId.isValid(data.engagementId) && ObjectId.isValid(data._id);
+    // Validate Data
+    var validData = await testPlanSchema.validate(data, { abortEarly: false });
   } catch (err) {
-    return { statusCode: 500, message: "Something went wrong with Yup Validation, check the Schema", info: valid, error: err }
+    return { statusCode: 422, message: "Yup Validation Failed", errors: err.errors }
   }
-  if (valid && validObjectIds) {
-    const validData = testPlanSchema.cast(data);
+
+  try {
+    // Connect to the Database
+    var db = await connectToDb();
+  } catch (err) {
+    console.log("Unable to connect to MongoDB")
+    return { statusCode: 500, message: "Unable to connect to MongoDB Server", errorName: err.name, error: err.message }
+  }
+  // Check that all Id strings are Valid Mongo Object Ids
+  var validDevices = !data.summaryBOM.map((dev) => ObjectId.isValid(dev.deviceId)).includes(false);
```

Comment:
User won't be able to edit devices or test cases of a test plan directly from this endpoint anyways, so I'm not sure why we need to check whether device Ids or test case ids are valid. They should have been already checked when they were added to the test plan?

Reply:
From the app's interface, they wouldn't be able to. They would however, be able to send the request by other means (like postman). and change the ids. This would be pretty unlikely, but possible nonetheless.